### PR TITLE
ensure `flux.job.JobID()` raises exceptions for all invalid jobids

### PR DESCRIPTION
**Pull request by grondo**
Problem: For some sets of invalid jobids, the `JobID()` constructor returns `0` (`ƒ1`) because exceptions are only raised if `fluid_parse()` returns `< 0` *and* `errno != 0`.

This PR fixes all the cases where `fluid_parse()` returns `-1` but leaves `errno` unset which leads to the above problem condition. Tests are added for `JobID()`, `fluid_parse()`, and `flux_jobid_parse()` to test expected behavior.

### CODE REVIEW

**Comment by grondo** [comment]
Pushed a commit to fix the dead code caught by CodeQL checks.

File: src/common/libutil/fluid.c [hunk 1]
```diff
@@ -291,14 +291,25 @@ static int fluid_decode_dothex (const char *s, fluid_t *fluid)
...
+            return -1;
+        }
+        if (i < 3 && *endptr != '.') {
+            errno = EINVAL;
             return -1;
-        if (i == 3 && *endptr != '\0')
+        }
+        if (i == 3 && *endptr != '\0') {
+            errno = EINVAL;
             return -1;
```
Comment (garlick) [inline src/common/libutil/fluid.c]: The pointer arithmetic hurts my brain.  What about something like this instead?
```c
static int fluid_decode_dothex (const char *s, fluid_t *fluid)
{
    unsigned int b[4];
    if (sscanf(s, "%04x.%04x.%04x.%04x%*c", &b[0], &b[1], &b[2], &b[3]) != 4) {
        errno = EINVAL;
        return -1;
    }
    *fluid = ((uint64_t)b[0] << 48)
        | ((uint64_t)b[1] << 32)
        | ((uint64_t)b[2] << 16)
        | b[3];
    return 0;
}
```

**Review by garlick — APPROVED** [review]
LGTM.  One suggestion that you can feel free to ignore.  I think your proposal is fine.

File: src/common/libutil/fluid.c (hunk 1 above)
Comment (grondo) [reply src/common/libutil/fluid.c]: This was preexisting, but sure that looks much better!

Comment (grondo) [reply src/common/libutil/fluid.c]: Maybe a separate issue though?

Comment (grondo) [reply src/common/libutil/fluid.c]: After looking at this closer, the existing `strtoul` parser actually lets through some bad inputs.
Therefore, let's prefer a tweaked version of @garlick's proposal above, replacing `%*c` (which silently eats trailing input, so validates `0.0.0.0.0`) with this:
```c
static int fluid_decode_dothex (const char *s, fluid_t *fluid)
{
    unsigned int b[4];
    int n = 0;
    if (sscanf (s, "%4x.%4x.%4x.%4x%n", &b[0], &b[1], &b[2], &b[3], &n) != 4
        || s[n] != '\0') {
        errno = EINVAL;
        return -1;
    }
    *fluid = ((uint64_t)b[0] << 48)
        | ((uint64_t)b[1] << 32)
        | ((uint64_t)b[2] << 16)
        | b[3];
    return 0;
}
```

**Comment by grondo** [comment]
Ok, pushed the additional change provided by @garlick here and some new tests to catch more corner case inputs in the dothex parser. Thanks! @garlick you may want to review the last two commits.

File: src/common/libutil/fluid.c (hunk 1 above)
Comment (garlick) [reply src/common/libutil/fluid.c]: Sorry about that, I thought the return value would be 5 if %*c matched anything but I see the return value is defined as "number of items matched and assigned", so bzzt.

**Review by garlick — APPROVED** [review]
👍
